### ISSUE TITLE
[ENH] untangling terminology for "global" forecasting

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#6690]
@sktime/core-developers, @Xinyu-Wu-0000, @fnhirwa, I think we ought to untangle terminology for "global" in the code base and especially in user facing documentation and tutorial notebooks.

Currently, it is used for two different concepts:

* forecasting models fitted to panel data non-trivially (not by broadcast), as opposed to models applied to panel data but a single-series instance is fitted to each series individually. This is used in this way in "global reduction", for instance, see `make_reduction`. In the M5 paper, this corresponds to "cross-learning", in classical econ/stats literatuer this may be referred to as "panel forecasting"
* forecasting models pre-trained in `fit` on different instances, and then `predict` on new, other instances. An example for this is the new series of PR by @Xinyu-Wu-0000, in particular deep learning models where the separate training instances in `fit` are passed as training batches to the networks. In a deep learning context, this is referred to as pre-learning, for foundation models, this may map onto "fine-tuning", see https://github.com/sktime/sktime/issues/6580

Opinions, @benHeid, @Xinyu-Wu-0000, @fnhirwa?
Also tagging @julian-fong, @pranavvp16 , @shlok191, @Z-Fran due to related items.

**Comment by benHeid** [comment]
Mhm, I need some clarification here: 

Is your first point referring to that there is one model fitted on different instances but will predict always all instances? 

If this is correct, than the second case is a generalization in my opinion. It does not require anymore that all of the time series are forecasted and it assumes that the underlying global model generalizes well enough so that it is also applicable to other time series (that are somewhat related).  In this case, we might think about to change the GlobalReducers to GlobalForecaster.

**Comment by fkiraly** [comment]
> Is your first point referring to that there is one model fitted on different instances but will predict always all instances?

Yes, in particuar the set of fit instances and the set of predict instances are equal.

> If this is correct, than the second case is a generalization in my opinion.

Not formally, no, because in the second case fit and predict instances are disjoint.

We could make it that they do not need to be, but that would require the boilerplate to hande intersection cases in a consistent way

**Comment by benHeid** [comment]
It depends on the user's input if they are disjoint or?

**Comment by fkiraly** [comment]
> It depends on the user's input if they are disjoint or?

Yes, though currently I think even if indices intersect or repeat, all global models consider them disjoint.

**Comment by benHeid** [comment]
Mhm yes if you are passing y it assumes it. 
However if you are not passing y it should be the same @Xinyu-Wu-0000 what happens if you are not providing any y to predict?

**Comment by Xinyu-Wu-0000** [comment]
> what happens if you are not providing any y to predict?

`y` from `fit` will be used instead assuming `X` to `predict` is from the same instances as `y` from `fit`.

**Comment by benHeid** [comment]
This means, that if no y is passed they are considered to be not disjoint or?

**Comment by Xinyu-Wu-0000** [comment]
> This means, that if no y is passed they are considered to be not disjoint or?

They are considered to be disjoint in the underlying `pytorch-forecasting` models. However, from the api perspective, it assumes predicting on the same set of time series passed to `fit`.

**Comment by fkiraly** [comment]
I think we should be really really clear in the `sktime` sided API design about how the user passes the following cases of global or panel forecasting:

* train on time series with index 1, ..., N; then make forecasts for all the time series with index 1, ..., N - "panel forecasting"
* train on time series with index 1, ..., N; then make forecasts for M new time series with index N+1, ..., N+M
* as a special case, train on time series 1, ..., N; then make forecast for 1 new time series with index N+1
* all cases should support arbitrary indices imo for the mtypes that allow this
* it should be *crystal clear* from an API perspective, to the user, whether they are in the 1st or the 2nd case, if M=N.

Advanced cases, I would recommend to park due to difficulty, but worth keeping in mind for upwards compatibility of the design:
* "intersection case" - train on 1, ..., N, forecast K, ..., K+M, where K <= N; this includes the "subset" case where K+M <= N too
* forecasting horizons that differ by instance
* batch update vs stream update, potentially mapping on `update`. Currently `update` is stream update only. Should this be extended to allow a batch update, with new instances? What about combined batch/stream, where we observe new instances as well as new history for already seen instances?

Should we open an enhancement proposal about this API? I think it would be really necessary, as the devil will be in the detail. Collaboration between @Xinyu-Wu-0000, core devs and other mentees whose project this touches, with @Xinyu-Wu-0000 having the first go at it?

**Comment by Xinyu-Wu-0000** [comment]
> * train on time series with index 1, ..., N; then make forecasts for all the time series with index 1, ..., N - "panel forecasting"

For huggingface and neuralforecast interface, if the `broadcasting` is set to `False`, and no `y` is passed to `predict`, It's "panel forecasting".

> * train on time series with index 1, ..., N; then make forecasts for M new time series with index N+1, ..., N+M - "global forecasting" or "cross-learning"

For huggingface and neuralforecast interface, if the `broadcasting` is set to `False`, and `y` is passed to `predict`, It's "global forecasting" or "cross-learning".

However, the 1st case is actually a special case of the 2nd case for DL models. Do we have models which can do "panel forecasting" but not "global forecasting"? 

> * "intersection case" - train on 1, ..., N, forecast K, ..., K+M, where K <= N; this includes the "subset" case where K+M <= N too

If K+M>N, `broadcasting==False`, it's "global forecasting".
If K+M>N, `broadcasting==True`, I don't know what it means.
If K+M <= N, `broadcasting==False`, it's "panel forecasting" (but not implemented yet for models can't do "global forecasting")
If K+M <=N, `broadcasting==True`, it's not implemented yet.

> * forecasting horizons that differ by instance

If `broadcasting==True`, it could be implemented in base class.
If `broadcasting==False`, it should be handled by each forecaster, maybe we can implement a api in base class to pass a batch of forecasting horizons to each forecaster.


> * batch update vs stream update, potentially mapping on `update`. Currently `update` is stream update only. Should this be extended to allow a batch update, with new instances? What about combined batch/stream, where we observe new instances as well as new history for already seen instances?

By `stream update only`, you mean only new history for already seen instances but no new instances right? Maybe we can create a new function `extend` to add new instances. And then we can have a function `update_extend(y_update, X_update, y_extend, X_extend, update_params)` for combining batch/stream.


For all above, I think we can do the following jobs:

1. If we have models that can't do "global forecasting" but can do "panel forecasting", we implement predict on a subset of fitted instances without broadcasting. (Might need to just refit on the subset of instances)
2. For `broadcasting=True`, we extend the vecterization to predict on a subset of fitted instances.
3. For `broadcasting=False`, we extend the `BaseForecaster` to pass different fh to underlying forecasters.
4. For `broadcasting=True`, we extend the vecterization to fit/predict with different fh for each instances.
5. New function `extend` and `update_extend`.

**Comment by yarnabrina** [comment]
@Xinyu-Wu-0000 will it be possible for you to share some code snippets for each of the different cases above, highlighting how as an user I'll achieve different above cases?

You can do it with pytorch forecasting or hugging face or neuralforecast or any other estimators.

**Comment by fkiraly** [comment]
agree, I would kindly request the same snippets, for explanation

**Comment by Xinyu-Wu-0000** [comment]
code snippet for "panel forecasting" and "global forecasting":

Set `y_to_predict` to true, it will be "global forecasting". Set `y_to_predict` to false, it will be "panel forecasting".

<details>

```py
import numpy as np

from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.neuralforecast import NeuralForecastLSTM, NeuralForecastRNN
from sktime.utils._testing.hierarchical import _make_hierarchical

broadcasting = False
y_to_predict = True
model_class = NeuralForecastRNN
max_prediction_length = 5
data = _make_hierarchical(
    hierarchy_levels=(5, 10),
    max_timepoints=20,
    min_timepoints=20,
    n_columns=3,
)

l1 = data.index.get_level_values(1).map(lambda x: int(x[3:]))
X_train = data.loc[l1 < 7, ("c0", "c1")]
y_train = data.loc[l1 < 7, "c2"].to_frame()
X_test = data.loc[l1 >= 7, ("c0", "c1")]
y_test = data.loc[l1 >= 7, "c2"].to_frame()

len_levels = len(data.index.names)
y_true = y_test if y_to_predict else y_train
y_true = y_true.groupby(level=list(range(len_levels - 1))).apply(
    lambda x: x.droplevel(list(range(len_levels - 1))).iloc[-max_prediction_length:]
)
X_futu = X_train.groupby(level=list(range(len_levels - 1))).apply(
    lambda x: x.droplevel(list(range(len_levels - 1))).iloc[-max_prediction_length:]
)
y_train = y_train.groupby(level=list(range(len_levels - 1))).apply(
    lambda x: x.droplevel(list(range(len_levels - 1))).iloc[:-max_prediction_length]
)
X_train = X_train.groupby(level=list(range(len_levels - 1))).apply(
    lambda x: x.droplevel(list(range(len_levels - 1))).iloc[:-max_prediction_length]
)
y_test = y_test.groupby(level=list(range(len_levels - 1))).apply(
    lambda x: x.droplevel(list(range(len_levels - 1))).iloc[:-max_prediction_length]
)

fh = ForecastingHorizon(
    np.arange(1, max_prediction_length + 1),
    is_relative=True,
    freq="D",
)

params = model_class.get_test_params()[2]
params["futr_exog_list"] = ["c0", "c1"]
params["broadcasting"] = broadcasting
model = model_class(**params)
model.fit(y=y_train, X=X_train, fh=fh)

y_pred = model.predict(
    X=X_test if y_to_predict else X_futu, y=y_test if y_to_predict else None
)
print(y_true)
print(y_pred)
```

</details>

**Comment by fkiraly** [comment]
Interesting!

Question: why do we need to pass `broadcasting` arg?

Could it not be inferred from whether `y` was passed to `predict`?

**Comment by Xinyu-Wu-0000** [comment]
> why do we need to pass `broadcasting` arg?

`broadcasing=True, y_to_predict=False` is how NeuralForecast and HuggingFace interface would work before the global forecasting api is added.
After  global forecasting api is added, we can also do `broadcasing=False, y_to_predict=True` and `broadcasing=False, y_to_predict=False`, so it's kind of back compatibility.

**Comment by benHeid** [comment]
In addition, I think it is also required to determine how the model should be fitted. Deciding only based on passed y in predict wouldn't work. Since this information is not available during fit.

**Comment by pranavvp16** [comment]
> In addition, I think it is also required to determine how the model should be fitted. Deciding only based on passed y in predict wouldn't work. Since this information is not available during fit.

`fit_strategy` can be passed in fit method to determine how the model is being is fitted. The values for `fit_strategy` could be `lora`, `finetune`, `pretrain` while default could be zero-shot forecasting.

**Comment by yarnabrina** [comment]
> > In addition, I think it is also required to determine how the model should be fitted. Deciding only based on passed y in predict wouldn't work. Since this information is not available during fit.
> 
> `fit_strategy` can be passed in fit method to determine how the model is being is fitted. The values for `fit_strategy` could be `lora`, `finetune`, `pretrain` while default could be zero-shot forecasting. But won't this be different for different models ?? and should be implementation specific

I am not sure if these fit strategies are applicable for all global forecasting models, e.g. for the `neuralforecast` ones. Please correct me if I am wrong.

**Comment by benHeid** [comment]
> `fit_strategy` can be passed in fit method to determine how the model is being is fitted. The values for `fit_strategy` could be `lora`, `finetune`, `pretrain` while default could be zero-shot forecasting. But won't this be different for different models ?? and should be implementation specific


Applying such PEFT methods is something different I would suppose. In global vs broadcasting it is simplified about if one model for all ts is fitted (global) or one model per time series (broadcast). A PEFT model can than also be applied with both strategies. Either fine tune a foundation model for each ts of a dataset separately (broadcasting) or fine tune it on the whole set of time series (global). 

Furthermore, I would strongly argue that deciding the strategy should be part of the init method since this is determining in sktime the model.

**Comment by benHeid** [comment]
> I am not sure if these fit strategies are applicable for all global forecasting models, e.g. for the `neuralforecast` ones. Please correct me if I am wrong.

Some of the PEFT methods such as LoRA are applicable to neural networks in general. Thus, in theory should be applicable to neuralforecast, if we are getting access to the internal torch model.

**Comment by fkiraly** [comment]
Related issue: https://github.com/sktime/sktime/issues/7332

Evidence that the lack of resolution to this issue is now causing user confusion... specifically the use in the tag vs use in the reduction forecasters. We should try to get to a resolution.

**Comment by joepvdbogaert** [comment]
Hey all, found this discussion after some confusion as well. I have a use case similar to M5: hierarchical sales forecasts for multiple brands, products, and sales channels. I've used [DirectTabularRegressionForecaster](https://www.sktime.net/en/stable/api_reference/auto_generated/sktime.forecasting.compose.DirectTabularRegressionForecaster.html) to train and evaluate a scikit-learn `HistGradientBoostingRegressor` on all combined data. Great so far.

Now I want to start deploying this model to an API, but struggle to find how I can forecast specific products on request, instead of exactly the set it is was trained on. To me it is very counterintuitive that a ML model cannot be fed with unseen data and produce a prediction. Seeing the term 'global forecasting' led to me to not even check if this was actually the case.

As context, some reasons I want to be able to predict on arbitrary data:
* I removed some outliers from the training data, but still need to product forecasts for them
* Product assortments and sales channels may change from one day to another
* I want to update the forecasts for specific products after their sales data is synced, which may be at different times and intervals
* I'd like to enable real-time updates to the forecasts after my users change certain input data (e.g. product availability, marketing campaigns, etc.)
* Predicting the complete dataset is slow, so forecasting everything and disregarding unused instances very very suboptimal

Is there any initiative going to make this work for reduction-based forecasters? I would be open to contribute if that helps. Think I have to bypass some of `sktime`'s useful tools now to make it work anyway.

**Comment by fkiraly** [comment]
Indeed, in the terminology of this issue, it means that reduction forecasters can make "global" forecasts in the sense of panel forecasts, but not cross-learning forecasts.

May I kindly ask that you open a new issue and request that feature? So we can track it as a feature request - this issue tracks a different problem, of terminology around "global".

**Comment by kdekker-kdr4** [comment]
@fkiraly Should this task still remain open or is the strategy set? I see improved clarifications in https://www.sktime.net/en/latest/examples/01c_forecasting_hierarchical_global.html#Global-forecasting/Panel-forecasting---introduction

**Comment by fkiraly** [comment]
That is an old notebook - @felipeangelimvieira is leading the rework of documentation currently.
